tests/test_java.py

```diff
@@ -275,36 +189,60 @@ def test_disable_java_profiling(application_pid, tmp_path, monkeypatch, caplog):
 def test_already_loaded_async_profiler_profiling_failure(tmp_path, monkeypatch, caplog, application_pid) -> None:
     with monkeypatch.context() as m:
         m.setattr("gprofiler.profilers.java.TEMPORARY_STORAGE_PATH", "/tmp/fake_gprofiler_tmp")
-        with JavaProfiler(
-            1,
-            5,
-            Event(),
-            str(tmp_path),
-            False,
-            True,
-            java_async_profiler_mode="cpu",
-            java_async_profiler_safemode=127,
-            java_async_profiler_args="",
-            java_safemode=JAVA_SAFEMODE_ALL,
-            java_mode="ap",
-        ) as profiler:
+        with make_java_profiler(str(tmp_path)) as profiler:
             profiler.snapshot()
 
-    with JavaProfiler(
-        1,
-        5,
-        Event(),
-        str(tmp_path),
-        False,
-        True,
-        java_async_profiler_mode="cpu",
-        java_async_profiler_safemode=127,
-        java_async_profiler_args="",
-        java_safemode=JAVA_SAFEMODE_ALL,
-        java_mode="ap",
-    ) as profiler:
+    with make_java_profiler(str(tmp_path)) as profiler:
         process = profiler._select_processes_to_profile()[0]
         assert any("/tmp/fake_gprofiler_tmp" in mmap.path for mmap in process.memory_maps())
         collapsed = snapshot_one_collaped(profiler)
         assert collapsed == Counter({"java;[Profiling skipped: async-profiler is already loaded]": 1})
         assert "Non-gProfiler async-profiler is already loaded to the target process" in caplog.text
+
+
+# test only once; and don't test in container - as it will go down once we kill the Java app.
+@pytest.mark.parametrize("in_container", [False])
+@pytest.mark.parametrize("check_app_exited", [False])  # we're killing it, the exit check will raise.
+def test_async_profiler_output_written_upon_jvm_exit(tmp_path, application_pid, assert_collapsed, caplog) -> None:
+    """
+    Make sure async-profiler writes output upon process exit (and we manage to read it correctly)
+    """
+    caplog.set_level(logging.DEBUG)
+
+    with make_java_profiler(str(tmp_path), duration=10) as profiler:
+
+        def delayed_kill():
+            time.sleep(3)
+            os.kill(application_pid, signal.SIGINT)
+
+        threading.Thread(target=delayed_kill).start()
+
+        process_collapsed = snapshot_one_collaped(profiler)
+        assert_collapsed(process_collapsed, check_comm=True)
```

Comment:
What does `check_comm=True` do? Because it's unused in `assert_function_in_collapsed()` 🙃

Reply:
That's a great question. Will check now in the history..

Reply:
Okay, I found https://github.com/Granulate/gprofiler/pull/89/commits/3fb545d0bf40f038a8e6d7e297268a44b35f0aa4 which added it, and then I found https://github.com/Granulate/gprofiler/pull/89/commits/c5ad646cec81734c1fe026ec9e39b50d633c3bfc which moved the code but actually removed the test.

I'm removing this entire mechanism as it's not important anymore (and won't work as I originally wrote it - it's not correct for all runtimes - e.g php - and not correct in general because the first stack is the container name).